src/realm/object-store/sync/app.cpp

```diff
@@ -723,85 +744,180 @@ std::string App::url_for_path(const std::string& path = "") const
-                m_app_route = m_base_route + this_app_path;
-                m_auth_route = m_app_route + auth_path;
-                m_sync_manager->set_sync_route(ws_hostname + base_path + this_app_path + sync_path);
-            }
-            catch (const AppError&) {
-                return completion(std::move(response));
-            }
+    m_config.transport->send_request_to_server(std::move(req), [this, completion = std::move(completion)](
+                                                                   const Request&, const Response& response) {
+        // If the response contains an error, then pass it up
+        if (response.http_status_code >= 300 || (response.http_status_code < 200 && response.http_status_code != 0)) {
+            return completion(std::move(response)); // early return
+        }
 
-            completion(util::none);
-        });
+        try {
+            auto json = parse<BsonDocument>(response.body);
+            auto hostname = get<std::string>(json, "hostname");
+            auto ws_hostname = get<std::string>(json, "ws_hostname");
+            auto deployment_model = get<std::string>(json, "deployment_model");
+            auto location = get<std::string>(json, "location");
+            m_sync_manager->perform_metadata_update([&](SyncMetadataManager& manager) {
+                manager.set_app_metadata(deployment_model, location, hostname, ws_hostname);
+            });
+
+            update_hostname(m_sync_manager->app_metadata());
+        }
+        catch (const AppError&) {
+            // Pass the response back to completion
+            return completion(std::move(response));
+        }
+        completion(util::none);
+    });
 }
 
-void App::post(std::string&& route, util::UniqueFunction<void(util::Optional<AppError>)>&& completion,
-               const BsonDocument& body)
+void App::post(std::string&& route, UniqueFunction<void(Optional<AppError>)>&& completion, const BsonDocument& body)
 {
     do_request(Request{HttpMethod::post, std::move(route), m_request_timeout_ms, get_request_headers(),
                        Bson(body).to_string()},
                handle_default_response(std::move(completion)));
 }
 
+void App::update_metadata_and_resend(Request&& request, UniqueFunction<void(const Response&)>&& completion,
+                                     const Optional<std::string>& new_hostname)
+{
+    // if we do not have metadata yet, we need to initialize it and send the
+    // request once that's complete; or if a new_hostname is provided, re-initialize
+    // the metadata with the updated location info
+    init_app_metadata(
+        [completion = std::move(completion), request = std::move(request), base_url = m_base_url,
+         anchor = shared_from_this()](const util::Optional<Response>& response) mutable {
+            if (response) {
+                if (response->http_status_code == static_cast<int>(realm::util::HTTPStatus::MovedPermanently)) {
```

Comment:
I think you want to cast http_status_code to HTTPStatus here, not the other way around.